slave/buildslave/runprocess.py

```diff
@@ -821,82 +822,105 @@ def kill(self, msg):
-            elif self.process.pgid is None:
-                log.msg("self.process has no pgid")
-            else:
-                log.msg("trying to kill process group %d" %
-                                                (self.process.pgid,))
+
+        def logAndSend(s):
+            log.msg(s)
+            self.sendStatus({'header': "\n" + s + "\n"})
+
+        msg += ", attempting to kill process with pid {} / group id {}".format(
+            getattr(self.process, 'pid', '<UNAVAILABLE>'),
+            getattr(self.process, 'gpid', '<UNAVAILABLE>')
+        )
+        logAndSend(msg)
+
+        try:
+            # let the PP know that we are killing it, so that it can ensure that
+            # the exit status comes out right
+            self.pp.killed = True
+
+            # keep track of whether we believe we've successfully killed something
+            hit = 0
+
+            # try signalling the process group
+            if not hit and self.useProcGroup and runtime.platformType == "posix":
+                sig = getattr(signal, "SIG"+ self.interruptSignal, None)
+
+                if sig is None:
+                    log.msg("signal module is missing SIG%s" % self.interruptSignal)
+                elif not hasattr(os, "kill"):
+                    log.msg("os module is missing the 'kill' function")
+                elif self.process.pgid is None:
+                    log.msg("self.process has no pgid")
+                else:
+                    logAndSend("trying to kill process group %d" %
+                                                    (self.process.pgid,))
+                    try:
+                        os.kill(-self.process.pgid, sig)
+                        logAndSend(" signal %s sent successfully" % sig)
+                        self.process.pgid = None
+                        hit = 1
+                    except OSError:
+                        logAndSend('failed to kill process group (ignored): %s' %
+                                (sys.exc_info()[1],))
+                        # probably no-such-process, maybe because there is no process
+                        # group
+                        pass
+
+            elif runtime.platformType == "win32":
+                if self.interruptSignal == None:
+                    log.msg("self.interruptSignal==None, only pretending to kill child")
+                elif self.process.pid is not None:
+                    logAndSend("using TASKKILL /F PID /T to kill pid %s" % self.process.pid)
+                    for i in range(5):
+                        try:
+                            subprocess.check_call("TASKKILL /F /PID %s /T" % self.process.pid)
+                            logAndSend("taskkill'd pid %s" % self.process.pid)
+                            hit = 1
+                            break
+                        except subprocess.CalledProcessError, e:
+                            if e.returncode == 128:
```

Comment:
maybe we should set `hit = 1` because we don't know proceed next `if not hit` if "Process no longer exists"

Reply:
I think you guys would do a better job at maintaining/finishing this, so please make whatever changes you want :)

And yea, it should definitely have a `hit = 1` here